### PR TITLE
Bulk loads

### PR DESCRIPTION
**Pull request by jonjove**
Work-in-progress bulk load implementation.

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

**Comment by vogel** [comment]
I think you forgot to add some files, like SqliteUtils.h?

**Comment by jonjove** [comment]
@vogel Thank you for noticing, they got lost in the rebase. Fixed.

File: src/ledger/LedgerTxnAccountSQL.cpp [hunk 1]
```diff
@@ -406,4 +407,133 @@ LedgerTxnRoot::Impl::encodeHomeDomainsBase64()
...
+    std::vector<LedgerEntry> entries;
+    if (mDatabase.isSqlite())
+    {
+        entries = sqliteSpecificBulkLoadAccounts(mDatabase, accountIDs);
+    }
+    else
+    {
+        std::abort();
+    }
+
```
Comment (marta-lokhova) [inline src/ledger/LedgerTxnAccountSQL.cpp]: The next 15 lines are duplicated in `LedgerTxnRoot::Impl::bulkLoadData`,` LedgerTxnRoot::Impl::bulkLoadOffers`, `LedgerTxnRoot::Impl::bulkLoadTrustLines` and should be extracted into a free util function, e.g.
```
std::unordered_map<LedgerKey, std::shared_ptr<LedgerEntry const>>
populateLoadedEntries(std::vector<LedgerKey> const& keys, std::vector<LedgerEntry> const& entries);
```

File: src/ledger/LedgerTxnImpl.h [hunk 2]
```diff
@@ -343,6 +349,15 @@ class LedgerTxnRoot::Impl
     getFromBestOffersCache(Asset const& buying, Asset const& selling,
                            BestOffersCacheEntry& defaultValue) const;
 
+    std::unordered_map<LedgerKey, std::shared_ptr<LedgerEntry const>>
```
Comment (marta-lokhova) [inline src/ledger/LedgerTxnImpl.h]: Should probably declare these four methods const.

File: src/ledger/LedgerTxn.cpp [hunk 3]
```diff
@@ -21,6 +21,27 @@
...
+                    std::make_shared<LedgerEntry const>(le));
+    }
+    for (auto const& key : keys)
+    {
+        if (res.find(key) == res.end())
+        {
+            res.emplace(key, nullptr);
+        }
+    }
+    assert(res.size() == keys.size());
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/ledger/LedgerTxn.cpp]: Changed `keys` to `std::unordered_set`. Added comment explicitly explaining the precondition. Added asserts that verify the precondition is satisfied.

File: src/ledger/LedgerTxnOfferSQL.cpp [hunk 4]
```diff
@@ -757,4 +758,178 @@ LedgerTxnRoot::Impl::dropOffers()
...
+        mSellerIDs.reserve(keys.size());
+        mOfferIDs.reserve(keys.size());
+        for (auto const& k : keys)
+        {
+            assert(k.type() == OFFER);
+            mSellerIDs.emplace_back(k.offer().sellerID);
+            mOfferIDs.emplace_back(unsignedToSigned(k.offer().offerID));
+        }
+
+        assert(mSellerIDs.size() == mOfferIDs.size());
```
[the remark this answers is not in the document]
Comment (jonjove) [reply src/ledger/LedgerTxnOfferSQL.cpp]: Removed.

**Comment by graydon** [comment]
r+ a6a7296

**Comment by graydon** [comment]
r+ 0e5dd39

**Pull request merged**